### PR TITLE
fix: hash the password

### PR DESCRIPTION
**Pull request by jannyHou**
## Description 

Fix the accesstoken generation function by checking the password instead of sending the plain password.

A follow up PR for https://github.com/strongloop/loopback4-example-shopping/pull/26
And connect to the first acceptance criteria in https://github.com/strongloop/loopback4-example-shopping/issues/35

- [x] Local test passes

### CODE REVIEW

File: src/utils/user.authentication.ts [hunk 1]
```diff
@@ -18,11 +20,21 @@ export async function getAccessTokenForUser(
   userRepository: UserRepository,
   credentials: Credentials,
 ): Promise<string> {
+  console.log('getAccessTokenForUser', credentials.password);
```
Comment (raymondfeng) [inline src/utils/user.authentication.ts]: It should be removed.

File: test/acceptance/user.controller.acceptance.ts [hunk 2]
```diff
@@ -135,36 +138,40 @@ describe('UserController', () => {
...
 
   describe('authentication functions', () => {
-    // TODO: fix storing the plain password in the following issue:
-    // https://github.com/strongloop/loopback-next/issues/1996
+    let plainPassword: string;
+
+    before('create new user', async () => {
+      plainPassword = user.password;
+      // Salt + Hash Password
+      user.password = await hashAsync(user.password, 10);
```
Comment (raymondfeng) [inline test/acceptance/user.controller.acceptance.ts]: Should it be the responsibility of the controller/repository to hash passwords?

File: src/utils/user.authentication.ts [hunk 3]
```diff
@@ -18,11 +20,21 @@ export async function getAccessTokenForUser(
...
   const foundUser = await userRepository.findOne({
-    where: {email: credentials.email, password: credentials.password},
+    where: {email: credentials.email},
   });
   if (!foundUser) {
-    throw new HttpErrors.Unauthorized('Wrong credentials!');
+    throw new HttpErrors.Unauthorized(
```
Comment (b-admike) [inline src/utils/user.authentication.ts]: Should we throw a 404 here or would that be a security risk?

File: src/utils/user.authentication.ts [hunk 4]
```diff
@@ -18,11 +20,21 @@ export async function getAccessTokenForUser(
...
+      `User with email ${credentials.email} not found`,
+    );
+  }
+  const passwordMatched = await compareAsync(
+    credentials.password,
+    foundUser.password,
+  );
+  if (!passwordMatched) {
+    throw new HttpErrors.Unauthorized('The credential is not correct!');
```
Comment (b-admike) [inline src/utils/user.authentication.ts]: nitpick: I would rephrase this to `The credentials are not correct`.

File: test/acceptance/user.controller.acceptance.ts (hunk 2 above)
Comment (b-admike) [reply test/acceptance/user.controller.acceptance.ts]: @jannyHou and I were talking about a hash utility function which can be called from the user repo or controller. I think it makes sense for the repository to call it when storing user passwords in the backend, but I see how that can be delegated to a controller too.

Comment (jannyHou) [reply test/acceptance/user.controller.acceptance.ts]: @raymondfeng @b-admike My opinion is unless we define a specific repository interface for model User and add a function for the hashing purpose, I tend to put the logic in controller. And as @b-admike said, we can extract it into utility/service that can be called from the controller.
Like what we have in [authentication utils](https://github.com/strongloop/loopback4-example-shopping/blob/master/src/utils/user.authentication.ts).

Comment (jannyHou) [reply test/acceptance/user.controller.acceptance.ts]: And I plan to refactor them into the util in the next PR, this one focus on fixing the plain password.

Explained in the PR description :)

A follow up PR for #26
And connect to the first acceptance criteria in #35 (there will be another PR after it to refactor the hash utils into a proper place.)

File: src/utils/user.authentication.ts (hunk 3 above)
Comment (jannyHou) [reply src/utils/user.authentication.ts]: @b-admike Good point, I think 404 makes more sense.

File: src/utils/user.authentication.ts [hunk 5]
```diff
@@ -20,21 +20,21 @@ export async function getAccessTokenForUser(
   userRepository: UserRepository,
   credentials: Credentials,
 ): Promise<string> {
-  console.log('getAccessTokenForUser', credentials.password);
+  console.log(credentials.email);
```
Comment (b-admike) [inline src/utils/user.authentication.ts]: Can we remove the `console.log` altogether here?

Comment (jannyHou) [reply src/utils/user.authentication.ts]: good catch removed and squashed into one commit.

File: test/acceptance/user.controller.acceptance.ts (hunk 2 above)
Comment (bajtos) [reply test/acceptance/user.controller.acceptance.ts]: @jannyHou Please do extract this code into utility/service that can be called from multiple places. For example tests.

File: test/unit/utils.authentication.unit.ts [hunk 6]
```diff
@@ -40,15 +76,19 @@ describe('authentication', () => {
...
-      expect('throws error').to.be.true();
-    } catch (err) {
-      expect(err.message).to.equal('jwt malformed');
-    }
+    const error = new JsonWebTokenError('jwt malformed');
+    return expect(decodeAccessToken(token, SECRET)).to.be.rejectedWith(error);
   });
+
+  async function createUser() {
+    user.password = await hashAsync(user.password, 10);
```
Comment (bajtos) [inline test/unit/utils.authentication.unit.ts]: @jannyHou 
In the test suite, please use the smallest number that's still supported by bcrypt - we use `4` in LoopBack 3.x

The higher the number is, the longer it takes to compute the hash. Slow hash computation is good in production as it makes brute-force attacks difficult, but also impractical in tests because it can significantly slow down the test suite. We have been through this in LoopBack 3.x, let's not repeat that mistake again.

Comment (jannyHou) [reply test/unit/utils.authentication.unit.ts]: @bajtos Thanks for sharing the experience and explain the reason behind using the smallest number! 

>In the test suite, please use the smallest number that's still supported by bcrypt 

Will do it in the next PR that extracts hash functions into utils.

Comment (jannyHou) [reply test/unit/utils.authentication.unit.ts]: @bajtos fixed in https://github.com/strongloop/loopback4-example-shopping/pull/41/files#diff-40bd51c7c6fe0f144ef2295401976579R20

File: test/acceptance/user.controller.acceptance.ts (hunk 2 above)
Comment (jannyHou) [reply test/acceptance/user.controller.acceptance.ts]: @bajtos See the refactor PR https://github.com/strongloop/loopback4-example-shopping/pull/41
